File: tests/node/test_secret_stores.js
/**
 * Integration tests for OpenLLM Node.js bindings (Rust via napi-rs)
 * 
 * Run with: npm test
 */

const assert = require('assert');
const { 
  EnvSecretStore, MemorySecretStore, listSecretStores,
  // Message types
  createSystemMessage, createUserMessage, createAssistantMessage,
  createToolResult, createToolError,
  listProviders,
  // Provider classes (note: NAPI uses camelCase for class names)
  OpenAiProvider, AnthropicProvider, GeminiProvider,
  OllamaProvider, MistralProvider, AzureOpenAiProvider, OpenRouterProvider,
  // Dynamic streaming
  streamChatWithProvider,
} = require('@openllm/native');

// Alias for consistency
const OpenAIProvider = OpenAiProvider;
const AzureOpenAIProvider = AzureOpenAiProvider;

// Test helpers
let passed = 0;
let failed = 0;

async function test(name, fn) {
  try {
    await fn();
    console.log(`  ✓ ${name}`);
    passed++;
  } catch (err) {
    console.log(`  ✗ ${name}`);
    console.log(`    Error: ${err.message}`);
    failed++;
  }
}

function assertEqual(actual, expected, msg = '') {
  if (actual !== expected) {
    throw new Error(`${msg} Expected ${JSON.stringify(expected)}, got ${JSON.stringify(actual)}`);
  }
}

function assertTrue(value, msg = '') {
  if (!value) {
    throw new Error(`${msg} Expected truthy value, got ${value}`);
  }
}

function assertFalse(value, msg = '') {
  if (value) {
    throw new Error(`${msg} Expected falsy value, got ${value}`);
  }
}

// =============================================================================
// Tests
// =============================================================================

async function testRegistry() {
  console.log('\n📋 Registry Tests');
  
  await test('listSecretStores returns built-in stores', () => {
    const stores = listSecretStores();
    assertTrue(stores.length >= 2, 'Should have at least 2 stores');
    
    const names = stores.map(s => s.name);
    assertTrue(names.includes('env'), 'Should include env store');
    assertTrue(names.includes('memory'), 'Should include memory store');
  });

  await test('store info has correct shape', () => {
    const stores = listSecretStores();
    const envStore = stores.find(s => s.name === 'env');
    
    assertTrue(typeof envStore.name === 'string');
    assertTrue(typeof envStore.description === 'string');
    assertTrue(typeof envStore.isPlugin === 'boolean');
    assertFalse(envStore.isPlugin, 'Built-in stores are not plugins');
  });
}

async function testEnvSecretStore() {
  console.log('\n🌍 EnvSecretStore Tests');
  
  const store = new EnvSecretStore();
  
  await test('has correct name', () => {
    assertEqual(store.name, 'env');
  });

  await test('is available', () => {
    assertTrue(store.isAvailable());
  });

  await test('reads direct env var', async () => {
    process.env.TEST_OPENLLM_SECRET = 'test-value-123';
    const value = await store.get('TEST_OPENLLM_SECRET');
    assertEqual(value, 'test-value-123');
    delete process.env.TEST_OPENLLM_SECRET;
  });

  await test('reads mapped provider name (openai)', async () => {
    process.env.OPENAI_API_KEY = 'sk-test-openai';
    const value = await store.get('openai');
    assertEqual(value, 'sk-test-openai');
    delete process.env.OPENAI_API_KEY;
  });

  await test('reads mapped provider name (anthropic)', async () => {
    process.env.ANTHROPIC_API_KEY = 'sk-ant-test';
    const value = await store.get('anthropic');
    assertEqual(value, 'sk-ant-test');
    delete process.env.ANTHROPIC_API_KEY;
  });

  await test('returns null for missing key', async () => {
    const value = await store.get('NONEXISTENT_KEY_12345');
    assertEqual(value, null);
  });

  await test('has() returns true for existing key', async () => {
    process.env.TEST_HAS_KEY = 'value';
    const result = await store.has('TEST_HAS_KEY');
    assertTrue(result);
    delete process.env.TEST_HAS_KEY;
  });

  await test('has() returns false for missing key', async () => {
    const result = await store.has('NONEXISTENT_KEY_12345');
    assertFalse(result);
  });

  await test('getInfo() returns correct info for existing key', async () => {
    process.env.TEST_INFO_KEY = 'value';
    const info = await store.getInfo('TEST_INFO_KEY');
    assertTrue(info.available);
    assertEqual(info.source, 'env');
    delete process.env.TEST_INFO_KEY;
  });

  await test('getInfo() returns not available for missing key', async () => {
    const info = await store.getInfo('NONEXISTENT_KEY_12345');
    assertFalse(info.available);
  });

  await test('store() throws ReadOnly error', async () => {
    try {
      await store.store('key', 'value');
      throw new Error('Should have thrown');
    } catch (err) {
      assertTrue(err.message.includes('read-only') || err.message.includes('ReadOnly'));
    }
  });

  await test('delete() throws ReadOnly error', async () => {
    try {
      await store.delete('key');
      throw new Error('Should have thrown');
    } catch (err) {
      assertTrue(err.message.includes('read-only') || err.message.includes('ReadOnly'));
    }
  });
}

async function testMemorySecretStore() {
  console.log('\n💾 MemorySecretStore Tests');
  
  const store = new MemorySecretStore();
  
  await test('has correct name', () => {
    assertEqual(store.name, 'memory');
  });

  await test('is available', () => {
    assertTrue(store.isAvailable());
  });

  await test('starts empty', () => {
    assertTrue(store.isEmpty());
    assertEqual(store.len(), 0);
  });

  await test('store and get', async () => {
    await store.store('key1', 'value1');
    const value = await store.get('key1');
    assertEqual(value, 'value1');
  });

  await test('has() returns true after store', async () => {
    await store.store('key2', 'value2');
    const result = await store.has('key2');
    assertTrue(result);
  });

  await test('len() increases after store', () => {
    assertTrue(store.len() >= 2);
  });

  await test('isEmpty() returns false after store', () => {
    assertFalse(store.isEmpty());
  });

  await test('getInfo() returns correct info', async () => {
    const info = await store.getInfo('key1');
    assertTrue(info.available);
    assertEqual(info.source, 'memory');
  });

  await test('delete removes key', async () => {
    await store.store('to_delete', 'value');
    assertTrue(await store.has('to_delete'));
    await store.delete('to_delete');
    assertFalse(await store.has('to_delete'));
  });

  await test('clear() removes all keys', () => {
    store.clear();
    assertTrue(store.isEmpty());
    assertEqual(store.len(), 0);
  });

  await test('update existing key', async () => {
    await store.store('update_key', 'original');
    assertEqual(await store.get('update_key'), 'original');
    await store.store('update_key', 'updated');
    assertEqual(await store.get('update_key'), 'updated');
  });
}

async function testMultipleStoreInstances() {
  console.log('\n🔀 Multiple Store Instances Tests');

  await test('memory stores are independent', async () => {
    const store1 = new MemorySecretStore();
    const store2 = new MemorySecretStore();
    
    await store1.store('key', 'value1');
    await store2.store('key', 'value2');
    
    assertEqual(await store1.get('key'), 'value1');
    assertEqual(await store2.get('key'), 'value2');
  });

  await test('env stores share same environment', async () => {
    const store1 = new EnvSecretStore();
    const store2 = new EnvSecretStore();
    
    process.env.SHARED_KEY = 'shared_value';
    
    assertEqual(await store1.get('SHARED_KEY'), 'shared_value');
    assertEqual(await store2.get('SHARED_KEY'), 'shared_value');
    
    delete process.env.SHARED_KEY;
  });
}

// =============================================================================
// Chat Message Tests
// =============================================================================

async function testChatMessages() {
  console.log('\n💬 ChatMessage Tests');

  await test('createSystemMessage', () => {
    const msg = createSystemMessage('You are helpful');
    assertEqual(msg.role, 'System');
    assertEqual(msg.content, 'You are helpful');
  });

  await test('createUserMessage', () => {
    const msg = createUserMessage('Hello');
    assertEqual(msg.role, 'User');
    assertEqual(msg.content, 'Hello');
  });

  await test('createAssistantMessage', () => {
    const msg = createAssistantMessage('Hi there!');
    assertEqual(msg.role, 'Assistant');
    assertEqual(msg.content, 'Hi there!');
  });
}

// =============================================================================
// Tool Tests
// =============================================================================

async function testToolTypes() {
  console.log('\n🔧 Tool Tests');

  await test('createToolResult', () => {
    const result = createToolResult('call_123', '72°F');
    assertEqual(result.callId, 'call_123');
    assertEqual(result.content, '72°F');
    assertFalse(result.isError);
  });

  await test('createToolError', () => {
    const result = createToolError('call_456', 'Not found');
    assertEqual(result.callId, 'call_456');
    assertTrue(result.isError);
  });
}

// =============================================================================
// Provider Tests
// =============================================================================

async function testProviders() {
  console.log('\n🏭 Provider Tests');

  await test('listProviders returns all providers', () => {
    const providers = listProviders();
    assertTrue(providers.length >= 7, 'Should have at least 7 providers');
    
    const ids = providers.map(p => p.id);
    assertTrue(ids.includes('openai'), 'Should include openai');
    assertTrue(ids.includes('anthropic'), 'Should include anthropic');
    assertTrue(ids.includes('gemini'), 'Should include gemini');
    assertTrue(ids.includes('ollama'), 'Should include ollama');
  });

  await test('provider metadata has correct structure', () => {
    const providers = listProviders();
    const openai = providers.find(p => p.id === 'openai');
    
    assertEqual(openai.displayName, 'OpenAI');
    assertTrue(openai.defaultApiBase.includes('openai.com'));
    assertTrue(openai.requiresApiKey);
    assertTrue(openai.defaultModels.length > 0);
  });

  await test('ollama does not require API key', () => {
    const providers = listProviders();
    const ollama = providers.find(p => p.id === 'ollama');
    
    assertFalse(ollama.requiresApiKey);
  });

  await test('provider has default models with capabilities', () => {
    const providers = listProviders();
    const openai = providers.find(p => p.id === 'openai');
    const model = openai.defaultModels[0];
    
    assertTrue(typeof model.id === 'string');
    assertTrue(typeof model.name === 'string');
    assertTrue(typeof model.contextLength === 'number');
    assertTrue(typeof model.capabilities.streaming === 'boolean');
  });
}

// =============================================================================
// Provider Class Tests
// =============================================================================

async function testProviderClasses() {
  console.log('\n🚀 Provider Class Tests');

  await test('OpenAIProvider can be instantiated', () => {
    const provider = new OpenAIProvider();
    assertEqual(provider.name, 'openai');
    const metadata = provider.metadata();
    assertEqual(metadata.id, 'openai');
    assertTrue(metadata.requiresApiKey);
  });

  await test('AnthropicProvider can be instantiated', () => {
    const provider = new AnthropicProvider();
    assertEqual(provider.name, 'anthropic');
    const metadata = provider.metadata();
    assertEqual(metadata.id, 'anthropic');
    assertTrue(metadata.requiresApiKey);
  });

  await test('GeminiProvider can be instantiated', () => {
    const provider = new GeminiProvider();
    assertEqual(provider.name, 'gemini');
    const metadata = provider.metadata();
    assertEqual(metadata.id, 'gemini');
  });

  await test('OllamaProvider can be instantiated', () => {
    const provider = new OllamaProvider();
    assertEqual(provider.name, 'ollama');
    const metadata = provider.metadata();
    assertEqual(metadata.id, 'ollama');
    assertFalse(metadata.requiresApiKey);
  });

  await test('MistralProvider can be instantiated', () => {
    const provider = new MistralProvider();
    assertEqual(provider.name, 'mistral');
  });

  await test('AzureOpenAIProvider can be instantiated', () => {
    const provider = new AzureOpenAIProvider();
    assertEqual(provider.name, 'azure');
  });

  await test('OpenRouterProvider can be instantiated', () => {
    const provider = new OpenRouterProvider();
    assertEqual(provider.name, 'openrouter');
  });

  await test('Provider has streamChat method', () => {
    const provider = new OpenAIProvider();
    assertTrue(typeof provider.streamChat === 'function');
  });

  await test('streamChatWithProvider function exists', () => {
    assertTrue(typeof streamChatWithProvider === 'function');
  });
}

// =============================================================================
// Main
// =============================================================================

async function main() {
  console.log('═══════════════════════════════════════════════════════════');
  console.log(' OpenLLM Node.js Bindings - Integration Tests');
  console.log('═══════════════════════════════════════════════════════════');

  await testRegistry();
  await testEnvSecretStore();
  await testMemorySecretStore();
  await testMultipleStoreInstances();
  await testChatMessages();
  await testToolTypes();
  await testProviders();
  await testProviderClasses();

  console.log('\n═══════════════════════════════════════════════════════════');
  console.log(` Results: ${passed} passed, ${failed} failed`);
  console.log('═══════════════════════════════════════════════════════════');

  if (failed > 0) {
    process.exit(1);
  }
}

main().catch(err => {
  console.error('Test runner error:', err);
  process.exit(1);
});

File: tests/python/test_secret_stores.py
#!/usr/bin/env python3
"""
Integration tests for OpenLLM Python bindings (Rust via PyO3)

Run with: pytest test_secret_stores.py -v
Or simply: python test_secret_stores.py
"""

import os
import sys

# Try to import pytest, fall back to simple runner
try:
    import pytest
    HAS_PYTEST = True
except ImportError:
    HAS_PYTEST = False

from openllm import (
    EnvSecretStore, MemorySecretStore, list_secret_stores,
    # New types
    ChatMessage, MessageRole, Tool, ToolCall, ToolResult,
    ModelConfig, ModelCapabilities, ProviderMetadata,
    list_providers,
)


# =============================================================================
# Registry Tests
# =============================================================================

class TestRegistry:
    """Tests for the secret store registry."""

    def test_list_returns_builtin_stores(self):
        """Registry should include built-in stores."""
        stores = list_secret_stores()
        assert len(stores) >= 2, "Should have at least 2 stores"
        
        names = [s.name for s in stores]
        assert "env" in names, "Should include env store"
        assert "memory" in names, "Should include memory store"

    def test_store_info_shape(self):
        """Store info should have correct attributes."""
        stores = list_secret_stores()
        env_store = next(s for s in stores if s.name == "env")
        
        assert isinstance(env_store.name, str)
        assert isinstance(env_store.description, str)
        assert isinstance(env_store.is_plugin, bool)
        assert env_store.is_plugin is False, "Built-in stores are not plugins"


# =============================================================================
# EnvSecretStore Tests
# =============================================================================

class TestEnvSecretStore:
    """Tests for environment variable secret store."""

    def setup_method(self):
        """Set up test fixtures."""
        self.store = EnvSecretStore()
        # Clean up any test env vars
        for key in list(os.environ.keys()):
            if key.startswith("TEST_OPENLLM_"):
                del os.environ[key]

    def teardown_method(self):
        """Clean up test fixtures."""
        for key in list(os.environ.keys()):
            if key.startswith("TEST_OPENLLM_"):
                del os.environ[key]

    def test_name(self):
        """Store should have correct name."""
        assert self.store.name == "env"

    def test_is_available(self):
        """Store should always be available."""
        assert self.store.is_available() is True

    def test_repr(self):
        """Store should have useful repr."""
        assert "EnvSecretStore" in repr(self.store)
        assert "env" in repr(self.store)

    def test_get_direct_env_var(self):
        """Should read direct environment variable."""
        os.environ["TEST_OPENLLM_SECRET"] = "test-value-123"
        assert self.store.get("TEST_OPENLLM_SECRET") == "test-value-123"

    def test_get_mapped_openai(self):
        """Should map 'openai' to OPENAI_API_KEY."""
        os.environ["OPENAI_API_KEY"] = "sk-test-openai"
        assert self.store.get("openai") == "sk-test-openai"
        del os.environ["OPENAI_API_KEY"]

    def test_get_mapped_anthropic(self):
        """Should map 'anthropic' to ANTHROPIC_API_KEY."""
        os.environ["ANTHROPIC_API_KEY"] = "sk-ant-test"
        assert self.store.get("anthropic") == "sk-ant-test"
        del os.environ["ANTHROPIC_API_KEY"]

    def test_get_missing_returns_none(self):
        """Should return None for missing key."""
        assert self.store.get("NONEXISTENT_KEY_12345") is None

    def test_has_existing_key(self):
        """has() should return True for existing key."""
        os.environ["TEST_OPENLLM_HAS"] = "value"
        assert self.store.has("TEST_OPENLLM_HAS") is True

    def test_has_missing_key(self):
        """has() should return False for missing key."""
        assert self.store.has("NONEXISTENT_KEY_12345") is False

    def test_get_info_existing(self):
        """get_info() should return correct info for existing key."""
        os.environ["TEST_OPENLLM_INFO"] = "value"
        info = self.store.get_info("TEST_OPENLLM_INFO")
        assert info.available is True
        assert info.source == "env"

    def test_get_info_missing(self):
        """get_info() should return not available for missing key."""
        info = self.store.get_info("NONEXISTENT_KEY_12345")
        assert info.available is False

    def test_store_raises_read_only(self):
        """store() should raise RuntimeError (read-only)."""
        with pytest.raises(RuntimeError) if HAS_PYTEST else self._raises(RuntimeError):
            self.store.store("key", "value")

    def test_delete_raises_read_only(self):
        """delete() should raise RuntimeError (read-only)."""
        with pytest.raises(RuntimeError) if HAS_PYTEST else self._raises(RuntimeError):
            self.store.delete("key")

    @staticmethod
    def _raises(exc_type):
        """Simple context manager for exception testing without pytest."""
        class RaisesContext:
            def __enter__(self):
                return self
            def __exit__(self, exc_type_actual, exc_val, exc_tb):
                if exc_type_actual is None:
                    raise AssertionError(f"Expected {exc_type.__name__} to be raised")
                if not issubclass(exc_type_actual, exc_type):
                    raise AssertionError(f"Expected {exc_type.__name__}, got {exc_type_actual.__name__}")
                return True  # Suppress the exception
        return RaisesContext()


# =============================================================================
# MemorySecretStore Tests
# =============================================================================

class TestMemorySecretStore:
    """Tests for in-memory secret store."""

    def setup_method(self):
        """Set up test fixtures."""
        self.store = MemorySecretStore()

    def test_name(self):
        """Store should have correct name."""
        assert self.store.name == "memory"

    def test_is_available(self):
        """Store should always be available."""
        assert self.store.is_available() is True

    def test_repr(self):
        """Store should have useful repr."""
        assert "MemorySecretStore" in repr(self.store)

    def test_starts_empty(self):
        """New store should be empty."""
        assert self.store.is_empty() is True
        assert len(self.store) == 0

    def test_store_and_get(self):
        """Should store and retrieve values."""
        self.store.store("key1", "value1")
        assert self.store.get("key1") == "value1"

    def test_has_after_store(self):
        """has() should return True after store."""
        self.store.store("key2", "value2")
        assert self.store.has("key2") is True

    def test_len_increases(self):
        """len() should increase after store."""
        self.store.store("a", "1")
        self.store.store("b", "2")
        assert len(self.store) == 2

    def test_is_empty_false_after_store(self):
        """is_empty() should return False after store."""
        self.store.store("key", "value")
        assert self.store.is_empty() is False

    def test_get_info_existing(self):
        """get_info() should return correct info."""
        self.store.store("key", "value")
        info = self.store.get_info("key")
        assert info.available is True
        assert info.source == "memory"

    def test_delete_removes_key(self):
        """delete() should remove the key."""
        self.store.store("to_delete", "value")
        assert self.store.has("to_delete") is True
        self.store.delete("to_delete")
        assert self.store.has("to_delete") is False

    def test_clear_removes_all(self):
        """clear() should remove all keys."""
        self.store.store("a", "1")
        self.store.store("b", "2")
        self.store.clear()
        assert self.store.is_empty() is True
        assert len(self.store) == 0

    def test_update_existing_key(self):
        """Storing to existing key should update."""
        self.store.store("key", "original")
        assert self.store.get("key") == "original"
        self.store.store("key", "updated")
        assert self.store.get("key") == "updated"


# =============================================================================
# Multiple Instances Tests
# =============================================================================

class TestMultipleInstances:
    """Tests for multiple store instances."""

    def test_memory_stores_independent(self):
        """Memory stores should be independent."""
        store1 = MemorySecretStore()
        store2 = MemorySecretStore()
        
        store1.store("key", "value1")
        store2.store("key", "value2")
        
        assert store1.get("key") == "value1"
        assert store2.get("key") == "value2"

    def test_env_stores_share_environment(self):
        """Env stores should share the same environment."""
        store1 = EnvSecretStore()
        store2 = EnvSecretStore()
        
        os.environ["TEST_SHARED_KEY"] = "shared_value"
        
        assert store1.get("TEST_SHARED_KEY") == "shared_value"
        assert store2.get("TEST_SHARED_KEY") == "shared_value"
        
        del os.environ["TEST_SHARED_KEY"]


# =============================================================================
# Chat Message Tests
# =============================================================================

class TestChatMessage:
    """Tests for ChatMessage type."""

    def test_system_message(self):
        """Can create system message."""
        msg = ChatMessage.system("You are helpful")
        assert msg.role == MessageRole.System
        assert msg.content == "You are helpful"

    def test_user_message(self):
        """Can create user message."""
        msg = ChatMessage.user("Hello")
        assert msg.role == MessageRole.User
        assert msg.content == "Hello"

    def test_assistant_message(self):
        """Can create assistant message."""
        msg = ChatMessage.assistant("Hi there!")
        assert msg.role == MessageRole.Assistant
        assert msg.content == "Hi there!"


# =============================================================================
# Tool Tests
# =============================================================================

class TestToolTypes:
    """Tests for tool-related types."""

    def test_tool_creation(self):
        """Can create a tool."""
        tool = Tool("get_weather", "Get the weather", None)
        assert tool.name == "get_weather"
        assert tool.description == "Get the weather"

    def test_tool_result_success(self):
        """Can create success result."""
        result = ToolResult.success("call_123", "72°F")
        assert result.call_id == "call_123"
        assert result.content == "72°F"
        assert result.is_error is False

    def test_tool_result_error(self):
        """Can create error result."""
        result = ToolResult.error("call_456", "Not found")
        assert result.call_id == "call_456"
        assert result.is_error is True


# =============================================================================
# Provider Tests
# =============================================================================

class TestProviders:
    """Tests for provider metadata."""

    def test_list_providers(self):
        """list_providers returns all providers."""
        providers = list_providers()
        assert len(providers) >= 7  # OpenAI, Anthropic, Gemini, Ollama, Mistral, Azure, OpenRouter
        
        names = [p.id for p in providers]
        assert "openai" in names
        assert "anthropic" in names
        assert "gemini" in names
        assert "ollama" in names

    def test_provider_metadata(self):
        """Provider metadata has correct structure."""
        providers = list_providers()
        openai = next(p for p in providers if p.id == "openai")
        
        assert openai.display_name == "OpenAI"
        assert "openai.com" in openai.default_api_base
        assert openai.requires_api_key is True

    def test_ollama_no_api_key(self):
        """Ollama doesn't require API key."""
        providers = list_providers()
        ollama = next(p for p in providers if p.id == "ollama")
        
        assert ollama.requires_api_key is False


# =============================================================================
# Model Config Tests
# =============================================================================

class TestModelConfig:
    """Tests for model configuration."""

    def test_model_config_creation(self):
        """Can create model config."""
        config = ModelConfig("gpt4", "openai", "gpt-4")
        assert config.id == "gpt4"
        assert config.provider == "openai"
        assert config.model == "gpt-4"

    def test_model_config_with_options(self):
        """Can create model config with options."""
        config = ModelConfig(
            id="gpt4",
            provider="openai",
            model="gpt-4",
            api_key="sk-test",
            api_base="https://custom.api",
            context_length=8192
        )
        assert config.api_key == "sk-test"
        assert config.api_base == "https://custom.api"
        assert config.context_length == 8192

    def test_model_capabilities(self):
        """Can create model capabilities."""
        caps = ModelCapabilities.full()
        assert caps.image_input is True
        assert caps.tool_calling is True
        assert caps.streaming is True


# =============================================================================
# Simple Test Runner (when pytest not available)
# =============================================================================

def run_simple_tests():
    """Run tests without pytest."""
    passed = 0
    failed = 0
    
    test_classes = [
        TestRegistry,
        TestEnvSecretStore,
        TestMemorySecretStore,
        TestMultipleInstances,
        TestChatMessage,
        TestToolTypes,
        TestProviders,
        TestModelConfig,
    ]
    
    for test_class in test_classes:
        print(f"\n{test_class.__name__}")
        instance = test_class()
        
        # Run setup if exists
        setup = getattr(instance, "setup_method", None)
        teardown = getattr(instance, "teardown_method", None)
        
        for name in dir(instance):
            if name.startswith("test_"):
                if setup:
                    setup()
                try:
                    getattr(instance, name)()
                    print(f"  ✓ {name}")
                    passed += 1
                except Exception as e:
                    print(f"  ✗ {name}")
                    print(f"    Error: {e}")
                    failed += 1
                finally:
                    if teardown:
                        teardown()
    
    print(f"\n{'═' * 60}")
    print(f" Results: {passed} passed, {failed} failed")
    print(f"{'═' * 60}")
    
    return failed == 0


# =============================================================================
# Main
# =============================================================================

if __name__ == "__main__":
    print("═" * 60)
    print(" OpenLLM Python Bindings - Integration Tests")
    print("═" * 60)
    
    if HAS_PYTEST and len(sys.argv) > 1 and sys.argv[1] != "--no-pytest":
        # Run with pytest
        sys.exit(pytest.main([__file__, "-v"] + sys.argv[1:]))
    else:
        # Run simple test runner
        success = run_simple_tests()
        sys.exit(0 if success else 1)
